cupy/cuda/compiler.py

```diff
@@ -276,6 +278,93 @@ def _hash_hexdigest(value):
 _hash_length = len(_hash_hexdigest(b''))  # 40 for SHA1
 
 
+class RestrictedUnpickler(pickle.Unpickler):
+    """An unpickler that only allow some safe classes."""
+
+    def find_class(self, module, name):
+        """Only allow a restricted set of classes."""
+        if module == __name__ and name == 'CachedCudaModule':
+            return CachedCudaModule
+        msg = "'{}.{}' is forbidden and cannot be unpickled."
+        raise pickle.UnpicklingError(msg.format(module, name))
+
+    @classmethod
+    def loads(cls, s):
+        """Helper function analogous to pickle.loads()."""
+        return cls(io.BytesIO(s)).load()
+
+
+class CachedCudaModule:
+    __slots__ = ['backend', 'cubin', 'mapping']
+
+    _pickle_protocol = 4  # first introduced in Python 3.4
```

Comment:
One more thing: Could you make the default to 5 for Python 3.8+? I guess we would likely drop Python 3.7 support for CuPy v11 as NumPy has dropped it, so it'd be nice to adopt the newer (and more performant) protocol.

But, to do so the value of `_pickle_protocol` must be added as part of the file hash to avoid issues when multiple Python environments exist locally.